File: C++/palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* rev = reverse(head);
        ListNode* a = head, *b = rev;
        for (; a && b; a = a->next, b = b->next)
        {
            if (a->val != b->val) {
                return false;
            }
        }
        
        return a == b;
    }
    
    ListNode* reverse(ListNode* head) {
        ListNode* result = nullptr;
        
        for (ListNode* n = head; n; n = n->next)
        {
            ListNode* node = new ListNode(n->val, result);
            result = node;
        }
        
        return result;
    }
};

File: Python3/palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        data = []
        node = head
        size = 0
        while node != None:
            data.append(node.val)
            node = node.next
            size += 1
        i = 0
        j = size - 1
        while i < j:
            if data[i] != data[j]:
                return False
            i += 1
            j -= 1
        return True
        